### ISSUE TITLE
Linux Retina support

### ISSUE THREAD

**Issue by andrey154** [factor/factor#1615]
When running factor UI (listener, browser) on Ubuntu 15.10 4K monitor there is no possibility to read UI. Everything is so small.

**Comment by mrjbq7** [comment]
We have good retina support on Mac OS X. Shouldn't be hard to fix that -- do you know how to detect 4K monitor using Gtk?

**Comment by andrey154** [comment]
Honestly I don't know...
To resolve the problem for now I changed 
basis/ui/text/pango/pango.factor:
! andreyf 72 dpi set-global
300 dpi set-global
And font size become acceptable. So probably it is neccessary to setup pango accordingly to screen dpi.
Also there are hardcoded initial sizes for listener and probably other apps.

I will try to answer your question later when i will get to know.

**Comment by andrey154** [comment]
It seems this is the answer http://stackoverflow.com/questions/417384/get-dpi-settings-via-gtk

**Comment by mrjbq7** [comment]
If you wouldn't mind, could you try copying this to the Listener and run it, tell me what it reports?

```factor
USE: gdk
LIBRARY: gdk
FUNCTION: gint gdk_window_get_scale_factor ( GdkWindow *window )

world get handle>> window>> gdk_window_get_scale_factor .
```

**Comment by mrjbq7** [comment]
Also, does this "work" for you?

```
2.0 gl-scale-factor set-global
```

Is the window too small? Or the contents too small?  Or both?

(I don't have a 4k monitor with Linux to test on, so I hope you don't mind if I ask you for a couple pieces of info to get this bug fixed)

**Comment by andrey154** [comment]
Pardon me I am new to factor. Your snippet gave this output:
Cannot resolve C library function
Library: DLL" libgdk-x11-2.0.so"
Symbol: gdk_window_get_scale_factor
DlError: /usr/lib/x86_64-linux-gnu/libgdk-x11-2.0.so: undefined symbol: gdk_window_get_scale_factor

Seems like the symbol is defined here:
rgrep gdk_window_get_scale_factor /usr/lib
...
Binary file /usr/lib/x86_64-linux-gnu/libgdk-3.so.0.1600.7 matches
...
"2.0 gl-scale-factor set-global" gave

resource:basis/ui/text/pango/pango.factor

34: 2.0 gl-scale-factor set-global
                       ^
No word named “gl-scale-factor” found in current vocabulary search path

Where to put it?

Sure, i can try to help. You are welcome to ask. But not today. It is deep night or early morning here :)

**Comment by andrey154** [comment]
There are both problems: small fonts and small windows.

**Comment by catb0t** [comment]
@andrey154 You can use triple-backticks for code, like this:

<pre><code>```
code goes here
```</code></pre>
Don't forget the newlines before and after the triple-backticks. See more [formatting help](https://guides.github.com/features/mastering-markdown/), if you're curious.

**Comment by andrey154** [comment]
This works: 
```
IN: scratchpad USE: gdk.ffi
IN: scratchpad  gdk_screen_get_default gdk_screen_get_resolution

--- Data stack:
216.0
```
What to do with gl-scale-factor? Where to put it?

**Comment by mrjbq7** [comment]
So, gl-scale-factor is used for Mac retina support from the OpenGL vocabulary:

http://docs.factorcode.org/content/word-gl-scale-factor,opengl.html

On the Mac, we had a correctly-sized window with incorrectly drawn (too small) content inside when retina support was turned on. 

On Linux, the HiDPI support might need to effectively double all the values which is fine, but slightly different than the Mac needed. 

So, just to clarify the issue is the window and its contents are both too small and your workaround was to set pango dpi and then resize the window larger? Did that effectively "fix" it for you?

**Comment by andrey154** [comment]
> So, just to clarify the issue is the window and its contents are both too small and your workaround was to set pango dpi and then resize the window larger? Did that effectively "fix" it for you?

Yes, this has "fixed" the issue. I mean both changes should be present pango dpi settings and apps window sizes. Icons are still very small but i don't care.
Though I know it is a hack :) I would try to propose a proper fix but i know factor only for couple of days.

You are welcome to try some patches on my machine! Just send diffs. Is there any more direct communication way?

**Comment by erg** [comment]
I have a 48" 4k monitor running at 3840x2160@60hz and factor looks normal sized. nvidia-binary driver 352.79 (the nouveau driver doesn't offer any resolution over 1900x...) I also had to switch away from Ubuntu because 16.04 broke my resolution for some reason. Debian has been great so far.

What's the difference with our setups? If I could reproduce your problem I bet I could fix it.

```
Debian linux, unstable (?)
Linux jasperbox 4.4.0-1-amd64 #1 SMP Debian 4.4.6-1 (2016-03-17) x86_64 GNU/Linux

nvidia driver 352.79
Nvidia GTX 970

Xfce 4.12


deb http://ftp.us.debian.org/debian/ stretch main non-free contrib
deb-src http://ftp.us.debian.org/debian/ stretch main non-free contrib

deb http://security.debian.org/debian-security stretch/updates main non-free contrib
deb-src http://security.debian.org/debian-security stretch/updates main non-free contrib
```

**Comment by andrey154** [comment]
My setup is:
```
Ubuntu 15.10 24'' 4k monitor at 3840x2160@60hz physical PPI 185.12 nvidia gtx 960 at nvidia proprietary driver
```
though gdk reports 216 ppi as i mentioned above.
Simply my monitor smaller than yours that is why you can see everything fine on you screen.
I think the problem could be more pronounced if somebody would use 5120x2880 @ 27'' Dell UP2715K which is having even more pixel density.
Every UI window is small and fonts are very small; there is no way to set them from any menu or config file. Also it is not adjusted automatically.

**Comment by erg** [comment]
Dell Precision 7760 with small Factor window and small font.

```
(base) ➜  factor git:(master) ✗ inxi -SPARM -GCDN -v1 -xGCRS

System:    Host: beefcake Kernel: 5.11.0-27-generic x86_64 bits: 64 compiler: N/A Desktop: Gnome 3.36.9 
           Distro: Ubuntu 20.04.3 LTS (Focal Fossa) 
Machine:   Type: Laptop System: Dell product: Precision 7760 v: N/A serial: <superuser/root required> 
           Mobo: Dell model: 0P4NHH v: A00 serial: <superuser/root required> UEFI: Dell v: 1.2.2 date: 06/29/2021 
CPU:       Topology: 8-Core model: Intel Xeon W-11955M bits: 64 type: MT MCP arch: N/A L2 cache: 24.0 MiB 
           flags: avx avx2 lm nx pae sse sse2 sse3 sse4_1 sse4_2 ssse3 vmx bogomips: 83558 
           Speed: 3201 MHz min/max: 800/4900 MHz Core speeds (MHz): 1: 3150 2: 1492 3: 2132 4: 4900 5: 801 6: 1564 7: 1738 
           8: 2453 9: 801 10: 799 11: 801 12: 801 13: 800 14: 893 15: 942 16: 976 
Graphics:  Device-1: Intel vendor: Dell driver: i915 v: kernel bus ID: 00:02.0 
           Device-2: NVIDIA vendor: Dell driver: nvidia v: 470.57.02 bus ID: 01:00.0 
           Display: x11 server: X.Org 1.20.11 driver: fbdev,modesetting,nouveau,vesa tty: N/A 
           OpenGL: renderer: NVIDIA RTX A5000 Laptop GPU/PCIe/SSE2 v: 4.6.0 NVIDIA 470.57.02 direct render: Yes 
Audio:     Device-1: Intel vendor: Dell driver: snd_hda_intel v: kernel bus ID: 00:1f.3 
           Device-2: NVIDIA driver: snd_hda_intel v: kernel bus ID: 01:00.1 
           Sound Server: ALSA v: k5.11.0-27-generic 
Network:   Device-1: Intel Ethernet I219-LM vendor: Dell driver: e1000e v: kernel port: efa0 bus ID: 00:1f.6 
           Device-2: Intel driver: iwlwifi v: kernel port: 3000 bus ID: 94:00.0 
Drives:    Local Storage: total: 1.82 TiB used: 80.62 GiB (4.3%) 
           ID-1: /dev/nvme0n1 vendor: Samsung model: SSD 980 PRO 2TB size: 1.82 TiB 
           ID-2: /dev/nvme1n1 model: KBG40ZNS256G NVMe KIOXIA 256GB size: 238.47 GiB 
RAID:      Message: No RAID data was found. 
Partition: ID-1: / size: 225.07 GiB used: 80.55 GiB (35.8%) fs: ext4 dev: /dev/nvme1n1p3 
Info:      Processes: 394 Uptime: 22m Memory: 125.03 GiB used: 2.23 GiB (1.8%) Init: systemd runlevel: 5 Compilers: gcc: 9.3.0 
           clang: 10.0.0-4ubuntu1 Shell: zsh v: 5.8 inxi: 3.0.38 
```


```
IN: scratchpad USE: gdk.ffi
IN: scratchpad  gdk_screen_get_default gdk_screen_get_resolution
192.0
```

**Comment by erg** [comment]
I have a machine that repros this bug.

![signal-2021-12-11-155750_001](https://user-images.githubusercontent.com/27430/146091150-ac138f16-7ee7-4d95-8815-b171809d7918.jpeg)

**Comment by erg** [comment]
Related to https://github.com/factor/factor/issues/2030, possibly we should close one of these as a dupe?

**Comment by mrjbq7** [comment]
Fixed in 0c6a126012b06a1e5f1a076ea2d3688c568a1bd9.
